TLM/TMPE.CitiesGameBridge/Service/NetService.cs

```diff
@@ -102,71 +102,108 @@ public class NetService : INetService {
             IterateNodeSegments(nodeId, ClockDirection.None, handler);
         }
 
-        public void IterateNodeSegments(ushort nodeId,
-                                        ClockDirection dir,
-                                        NetSegmentHandler handler) {
-            NetManager netManager = Singleton<NetManager>.instance;
+        /// <summary>
+        /// Given a <paramref name="nodeId"/>, retrieves the assocaited <see cref="NetNode"/>.
+        ///
+        /// A <c>nodeId</c> of <c>0</c> returns <c>false</c>.
+        /// </summary>
+        /// 
+        /// <param name="nodeId">The id of the segment to retrieve.</param>
+        /// <param name="node">The returned segment.</param>
+        /// 
+        /// <returns>Returns <c>true</c> if successful, otherwise <c>false</c>.</returns>
+        public bool TryGetNode(ushort nodeId, out NetNode node) {
+            if (nodeId == 0) {
+                node = default;
+                return false;
+            }
 
-            bool ProcessFun(ushort nId, ref NetNode node) {
-                if (dir == ClockDirection.None) {
-                    for (int i = 0; i < 8; ++i) {
-                        ushort segmentId = node.GetSegment(i);
-                        if (segmentId != 0) {
-                            if (!handler(
-                                    segmentId,
-                                    ref netManager.m_segments.m_buffer[segmentId])) {
-                                break;
-                            }
-                        }
-                    }
-                } else {
-                    ushort segmentId = 0;
-                    for (int i = 0; i < 8; ++i) {
-                        segmentId = node.GetSegment(i);
-                        if (segmentId != 0) {
-                            break;
-                        }
-                    }
-                    ushort initSegId = segmentId;
-
-                    while (true) {
-                        if (segmentId != 0) {
-                            if (!handler(
-                                    segmentId,
-                                    ref netManager.m_segments.m_buffer[segmentId])) {
-                                break;
-                            }
-                        }
+            node = Singleton<NetManager>.instance.m_nodes.m_buffer[nodeId];
```

Comment:
Not sure if it is within the scope of this review but maybe we can cache `Singleton<NetManager>.instance.m_nodes.m_buffer` as `NodeBuffer` during initialization to increase speed.

Reply:
Ideally I'd like to create some manager classes to help with that but sort of based on how dymanoid does things in Real Time mod.

He has these connection classes - [git repo](https://github.com/dymanoid/RealTime/tree/master/src/RealTime/GameConnection) - where he stores his library functions for interacting with vanilla.

And then he adds those in to his other classes - [example](https://github.com/dymanoid/RealTime/blob/master/src/RealTime/CustomAI/RealTimeBuildingAI.cs#L77). It makes subsequent code much more readable.

We kind of have something like that in TMPE, which also has managers, but it's a big heaving mess of intermingled code currently.